### PR TITLE
Option to reuse an existing IoT Hub

### PR DESCRIPTION
**Pull request by tiagoshibata**
# Type of change?

- [ ] Bug fix
- [X] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description

Add question to reuse an existing IoT Hub to the configuration prompts. The use case is an user with an existing IoT Hub (e.g. from their PCS1 deployment) who wants to connects the PCS2 portal to it.

The user can provide the connection string of his IoT Hub, in which case the iothub-manager microservice will use it, or an empty string, in which case a new IoT Hub is created.

This PR is incomplete and I'd like to get some help with ARM. I will comment on this PR where it needs some work.

**Checklist:**

- [X] All tests passed
- [X] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: remotemonitoring/armtemplates/basic-static-map.json [hunk 1]
```diff
@@ -584,7 +591,6 @@
             "location": "[variables('location')]",
             "dependsOn": [
                 "[resourceId('Microsoft.Compute/virtualMachines/extensions', parameters('vmName'), 'DockerExtension')]",
-                "[resourceId('Microsoft.Devices/Iothubs/', parameters('iotHubName'))]",
```
Comment (tiagoshibata) [inline remotemonitoring/armtemplates/basic-static-map.json]: This `dependsOn` entry needs to be conditional (the dependency doesn't exist if the Hub is already deployed outside this template). Is there a way conditionally have the list entry? I couldn't find anything in the [reference](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions).

File: remotemonitoring/armtemplates/basic-static-map.json [hunk 2]
```diff
@@ -598,7 +604,7 @@
...
..., parameters('microServiceRuntime'), ' ', parameters('microServiceVersion'), ' ',  parameters('aadTenantId'), ' ', parameters('aadClientId'), ' ', parameters('aadInstance'), ' \"', concat('HostName=', reference(variables('iotHubResourceId')).hostName, ';SharedAccessKeyName=', variables('iotHubKeyName'), ';SharedAccessKey=', listkeys(variables('iotHubKeyResource'), variables('iotHubApiVersion')).primaryKey), '\" \"', concat('AccountEndpoint=', reference(variables('documentDBResourceId')).documentEndpoint, ';AccountKey=', listkeys(variables('documentDBResourceId'), variables('documentDBApiVersion')).primaryMasterKey, ';'), '\" ', reference(variables('iotHubResourceId')).eventHubEndpoints.events.path, ' ', reference(variables('iotHubResourceId')).eventHubEndpoints.events.endpoint, ' ', reference(variables('iotHubResourceId')).eventHubEndpoints.events.partitionCount, ' ', parameters('storageName'), ' ', listKeys(variables('storageResourceId'), providers('Microsoft.Storage', 'storageAccounts').apiVersions[0]).keys[0].value, ' \"', parameters('storageEndpointSuffix'), '\" \"', parameters('remoteEndpointCertificate'), '\" \"', parameters('remoteEndpointCertificateKey'), '\" \"static\"')]"
```
Comment (tiagoshibata) [inline remotemonitoring/armtemplates/basic-static-map.json]: I cannot `reference` a resource from outside this template. What's the recommended way to go when I need to access parameters from external resources?

Right now I have hardcoded the partition count, path and endpoint for testing, but we need a way to read them dynamically.

File: remotemonitoring/armtemplates/basic.json [hunk 3]
```diff
@@ -608,7 +615,6 @@
             "location": "[variables('location')]",
             "dependsOn": [
                 "[resourceId('Microsoft.Compute/virtualMachines/extensions', parameters('vmName'), 'DockerExtension')]",
-                "[resourceId('Microsoft.Devices/Iothubs/', parameters('iotHubName'))]",
```
Comment (tiagoshibata) [inline remotemonitoring/armtemplates/basic.json]: Same comment about `dependsOn`.

File: remotemonitoring/armtemplates/basic.json [hunk 4]
```diff
@@ -623,7 +629,10 @@
...
...' ',  parameters('aadTenantId'), ' ', parameters('aadClientId'), ' ', parameters('aadInstance'), ' \"', concat('HostName=', reference(variables('iotHubResourceId')).hostName, ';SharedAccessKeyName=', variables('iotHubKeyName'), ';SharedAccessKey=', listkeys(variables('iotHubKeyResource'), variables('iotHubApiVersion')).primaryKey), '\" \"', concat('AccountEndpoint=', reference(variables('documentDBResourceId')).documentEndpoint, ';AccountKey=', listkeys(variables('documentDBResourceId'), variables('documentDBApiVersion')).primaryMasterKey, ';'), '\" ', reference(variables('iotHubResourceId')).eventHubEndpoints.events.path, ' ', reference(variables('iotHubResourceId')).eventHubEndpoints.events.endpoint, ' ', reference(variables('iotHubResourceId')).eventHubEndpoints.events.partitionCount, ' ', parameters('storageName'), ' ', listKeys(variables('storageResourceId'), providers('Microsoft.Storage', 'storageAccounts').apiVersions[0]).keys[0].value, ' \"', parameters('storageEndpointSuffix'), '\" \"', parameters('remoteEndpointCertificate'), '\" \"', parameters('remoteEndpointCertificateKey'), '\" \"', listSecrets(variables('bingMapsName'), variables('bingMapsVersion')).QueryKey, '\"')]"
```
Comment (tiagoshibata) [inline remotemonitoring/armtemplates/basic.json]: Same comment about `reference`.

File: tsconfig.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     /* Basic Options */
-    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', or 'ESNEXT'. */
+    "target": "es2017",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', or 'ESNEXT'. */
```
Comment (tiagoshibata) [inline tsconfig.json]: I moved to `es2017` for `Array.prototype.includes`. Let me know if there was a technical reason to stay on `es6`.

File: remotemonitoring/armtemplates/standard.json [hunk 6]
```diff
@@ -498,4 +501,4 @@
             "value": "[listSecrets(variables('bingMapsName'), variables('bingMapsVersion')).QueryKey]"
```
Comment (tiagoshibata) [inline remotemonitoring/armtemplates/standard.json]: There are a bunch of output values that `reference` the IoT Hub (e.g. [see line 473](https://github.com/Azure/pcs-cli/pull/239/files#diff-0d5a330fd8bf17136fefca28a8f04322R473)). These will need some workaround too.

**Comment by tiagoshibata** [comment]
Let me know what you think... We could have a separate ARM template for deployments reusing IoT Hubs (just like we have a separate one for static maps), but I'd rather not duplicate the whole template because of a few small differences.

File: remotemonitoring/armtemplates/basic-static-map.json (hunk 1 above)
Comment (ppathan) [reply remotemonitoring/armtemplates/basic-static-map.json]: @tiagoshibata Please take a look at this for conditional statements within ARM template. https://azure.microsoft.com/en-us/blog/create-flexible-arm-templates-using-conditions-and-logical-functions/

File: remotemonitoring/armtemplates/basic-static-map.json (hunk 2 above)
Comment (ppathan) [reply remotemonitoring/armtemplates/basic-static-map.json]: One of the suggestion i can think of is passing those existing values as parameters or try to use reference but with full path. If you use parameters to pass in existing IotHub values then you can use conditional statements to either use resources created within the template or the parameters being passed in. https://azure.microsoft.com/en-us/blog/create-flexible-arm-templates-using-conditions-and-logical-functions/

**Pull request closed without merging**